### PR TITLE
4.2/perms

### PR DESCRIPTION
**Pull request by weeryan17**
- [x] Completed PR - Has all features or fixes that it was meant to and isn't half done
- [x] Follows the style guide - Style guide found [here](https://github.com/FlareBot/FlareBot/blob/master/CONTRIBUTING.md)
- [x] All changes have been tested

## Added
clone and rename group options
This implements suggestion 65

### CODE REVIEW

File: src/main/java/stream/flarebot/flarebot/permissions/PerGuildPermissions.java [hunk 1]
```diff
@@ -113,4 +113,26 @@ public void moveGroup(Group group, int pos) {
         groups.remove(index);
         groups.add(pos, group);
     }
+
+    public boolean cloneGroup(Group group, String newGroupName) {
+        if (hasGroup(newGroupName)) {
+            return false;
+        } else {
+            Group newGroup = new Group(newGroupName);
+            for(String permission: group.getPermissions()) {
```
Comment (WalshyDev) [inline src/main/java/stream/flarebot/flarebot/permissions/PerGuildPermissions.java]: space before `:` also remove brackets, single line body

File: src/main/java/stream/flarebot/flarebot/permissions/PerGuildPermissions.java [hunk 2]
```diff
@@ -113,4 +113,26 @@ public void moveGroup(Group group, int pos) {
         groups.remove(index);
         groups.add(pos, group);
     }
+
+    public boolean cloneGroup(Group group, String newGroupName) {
+        if (hasGroup(newGroupName)) {
```
Comment (WalshyDev) [inline src/main/java/stream/flarebot/flarebot/permissions/PerGuildPermissions.java]: remove brackets, single line body

File: src/main/java/stream/flarebot/flarebot/permissions/PerGuildPermissions.java [hunk 3]
```diff
@@ -113,4 +113,26 @@ public void moveGroup(Group group, int pos) {
...
+            for(String permission: group.getPermissions()) {
+                newGroup.addPermission(permission);
+            }
+            groups.add(newGroup);
+            return true;
+        }
+    }
+
+    public boolean renameGroup(Group group, String name) {
+        if (hasGroup(name)) {
```
Comment (WalshyDev) [inline src/main/java/stream/flarebot/flarebot/permissions/PerGuildPermissions.java]: Same here

File: src/main/java/stream/flarebot/flarebot/commands/management/PermissionsCommand.java [hunk 4]
```diff
@@ -359,7 +375,9 @@ public String getUsage() {
                 "`{%}permissions group <group> list [page]` - lists the permissions this group has.\n" +
                 "`{%}permissions group <group> massadd <@everyone/@here/role>` - Puts everyone with the giving role into the group.\n" +
                 "`{%}permissions group <group> clear` - Removes all permissions from this group!\n" +
-                "`{%}permissions group <group> move <pos>` - Moves the group to a different position on the hierarchy.\n\n" +
+                "`{%}permissions group <group> move <pos>` - Moves the group to a different position on the hierarchy.\n" +
+                "`{%}permissions group <group> clone <new_group>` - clones a group\n" +
+                "`{%}permissions group <group> rename <new_name>` - renames the group\n\n" +
```
Comment (binaryoverload) [inline src/main/java/stream/flarebot/flarebot/commands/management/PermissionsCommand.java]: Start the descriptions with capitals please

File: src/main/java/stream/flarebot/flarebot/commands/management/PermissionsCommand.java [hunk 5]
```diff
@@ -148,6 +148,22 @@ public void onCommand(User sender, GuildWrapper guild, TextChannel channel, Mess
...
+                        } else {
+                            MessageUtils.sendWarningMessage("Error cloning group (The clone might already exist)", channel);
+                            return;
+                        }
+                    } else if (args[2].equalsIgnoreCase("rename") && args.length >= 4) {
+                        if(guild.getPermissions().renameGroup(group, args[3])) {
+                            MessageUtils.sendMessage("Renamed group Successfully", channel);
+                            return;
+                        } else {
+                            MessageUtils.sendWarningMessage("Error renaming group (The rename might overlap a group)", channel);
```
Comment (binaryoverload) [inline src/main/java/stream/flarebot/flarebot/commands/management/PermissionsCommand.java]: Change it to "The destination group might already exist"

File: src/main/java/stream/flarebot/flarebot/commands/management/PermissionsCommand.java (hunk 4 above)
Comment (WalshyDev) [reply src/main/java/stream/flarebot/flarebot/commands/management/PermissionsCommand.java]: Also add full stops

Comment (binaryoverload) [reply src/main/java/stream/flarebot/flarebot/commands/management/PermissionsCommand.java]: :+1:

File: src/main/java/stream/flarebot/flarebot/commands/management/PermissionsCommand.java [hunk 6]
```diff
@@ -148,6 +148,22 @@ public void onCommand(User sender, GuildWrapper guild, TextChannel channel, Mess
                             MessageUtils.sendSuccessMessage("Moved group `" + groupString + "` to position " + pos, channel, sender);
                             return;
                         }
+                    } else if (args[2].equalsIgnoreCase("clone") && args.length >= 4) {
+                        if(guild.getPermissions().cloneGroup(group, args[3])) {
+                            MessageUtils.sendMessage("Cloned group Successfully", channel);
+                            return;
+                        } else {
+                            MessageUtils.sendWarningMessage("Error cloning group (The clone might already exist)", channel);
```
Comment (WalshyDev) [inline src/main/java/stream/flarebot/flarebot/commands/management/PermissionsCommand.java]: Change this to the group might already exist

**Pull request merged**